File: transpiler/javatests/com/google/j2cl/readable/java/cast/Casts.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cast;

import java.io.Serializable;

public class Casts {

  public void testCastToClass() {
    Object o = new Object();
    Casts c = (Casts) o;
  }

  public void testCasToInterface() {
    Object o = new Object();
    Serializable s = (Serializable) o;
  }

  public void testCastToBoxedType() {
    Object o = new Integer(1);
    Byte b = (Byte) o;
    Double d = (Double) o;
    Float f = (Float) o;
    Integer i = (Integer) o;
    Long l = (Long) o;
    Short s = (Short) o;
    Number n = (Number) o;
    Character c = (Character) o;
    Boolean bool = (Boolean) o;
  }

  public void testCastToArray() {
    Object foo = (Object[]) null;
    Object bar = (Object[][]) null;
  }

  private void functionForInstanceofTest() {}

  public void testCastObjectAfterInstanceOf() {
    Object o = new Object();
    if (o instanceof Casts) {
      Casts c = (Casts) o;
    }
    if (o instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) o;
      o = new Object();
      Casts cNotAvoidCastsTo = (Casts) o;
    }
    if (o instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) o;
      o = new Foo();
      Casts cNotAvoidCastsTo = (Casts) o;
    }
    if (o instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) o;
      o = new Object();
      Casts cAvoidCastsTo = (Casts) o;
    }
    if (o instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) o;
      functionForInstanceofTest();
      Casts cNotAvoidCastsTo = (Casts) o;
    }
  }

  /** Class for testing purposes */
  private class Foo {
    public Object field = new Object();

    public Object method() {
      return new Object();
    }
  }

  public void testCastFieldAfterInstanceOf() {
    Foo foo = new Foo();
    if (foo.field instanceof Casts) {
      Casts c = (Casts) foo.field;
    }
    if (foo.field instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) foo.field;
      foo.field = new Object();
      Casts cNotAvoidCastsTo = (Casts) foo.field;
    }
    if (foo.field instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) foo.field;
      foo = new Foo();
      Casts cNotAvoidCastsTo = (Casts) foo.field;
    }
    if (foo.field instanceof Casts) {
      Casts cAvoidsCastsTo = (Casts) foo.field;
      functionForInstanceofTest();
      Casts cNotAvoidCastsTo = (Casts) foo.field;
    }
  }

  public void testCaseMethodAfterInstanceOf() {
    Foo foo = new Foo();
    if (foo.method() instanceof Casts) {
      Casts cNotAvoidCastsTo = (Casts) foo.method();
    }
  }

  public void testPrecedence() {
    Object foo = "foo";
    Object bar = "bar";
    Integer notString = 123;
    String s1 = (String) (false ? foo : bar);
    String s2 = (String) ("foo" + notString);
  }
}

File: transpiler/javatests/com/google/j2cl/readable/kotlin/cast/Casts.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package cast

import java.io.Serializable

class Casts {
  fun testCastToClass() {
    val o = Any()
    val c = o as Casts?
  }

  fun testCasToInterface() {
    val o = Any()
    val s = o as Serializable?
  }

  fun testCastToBoxedType() {
    val o: Any? = null
    val b: Byte? = o as Byte?
    val d: Double? = o as Double?
    val f: Float? = o as Float?
    val i: Int? = o as Int?
    val l: Long? = o as Long?
    val s: Short? = o as Short?
    val n: Number? = o as Number?
    val c: Char? = o as Char?
    val bool: Boolean? = o as Boolean?
  }

  fun testCastToArray() {
    val foo: Array<Any?>? = null as Array<Any?>?
    val bar: Array<Array<Any?>?>? = null as Array<Array<Any?>?>?
  }

  private fun functionForInstanceofTest() {}

  fun testCastVarialeAfterInstanceOf() {
    var o: Any? = Any()
    if (o is Casts) {
      val c: Casts = o
    }
    if (o is Casts) {
      val cAvoidsCastsTo: Casts = o
      o = Any()
      val cNotAvoidCastsTo: Casts? = o as Casts?
    }
    if (o is Casts) {
      val cAvoidsCastsTo: Casts = o
      o = Foo()
      val cNotAvoidCastsTo: Casts? = o as Casts?
    }
    if (o is Casts) {
      val cAvoidsCastsTo: Casts = o
      o = Any()
      val cAvoidCastsTo: Casts? = o as Casts?
    }
    if (o is Casts) {
      val cAvoidsCastsTo: Casts = o
      functionForInstanceofTest()
      val cNotAvoidCastsTo: Casts = o
    }

    val c = if (o is Casts) o else null
  }

  /** Class for testing purposes */
  private inner class Foo {
    var field: Any = Any()

    fun method(): Any {
      return Any()
    }
  }

  fun testCastFieldAfterInstanceOf() {
    var foo: Foo = Foo()
    if (foo.field is Casts) {
      val c = foo.field as Casts?
    }
    if (foo.field is Casts) {
      val cAvoidsCastsTo = foo.field as Casts?
      foo.field = Any()
      val cNotAvoidCastsTo = foo.field as Casts?
    }
    if (foo.field is Casts) {
      val cAvoidsCastsTo = foo.field as Casts?
      foo = Foo()
      val cNotAvoidCastsTo = foo.field as Casts?
    }
    if (foo.field is Casts) {
      val cAvoidsCastsTo = foo.field as Casts?
      functionForInstanceofTest()
      val cNotAvoidCastsTo = foo.field as Casts?
    }

    val c = if (foo.field is Casts) foo.field as Casts else null
  }

  fun testCaseMethodAfterInstanceOf() {
    val foo: Foo = Foo()
    if (foo.method() is Casts) {
      val cNotAvoidCastsTo = foo.method() as Casts?
    }

    val c = if (foo.method() is Casts) foo.method() as Casts else null
  }

  fun testCastNullability() {
    val a: Any? = Foo()
    val aa = a as Foo

    val b: Any? = Foo()
    val bb = b as Foo?

    val c: Any? = Foo()
    val cc = c as? Foo

    val d: Any? = Foo()
    val dd = d as? Foo?
  }

  fun testPrecedence() {
    val foo: Any = "foo"
    val bar: Any = "bar"
    val notString = 123
    val s1 = (if (false) foo else bar) as String
    val s2 = ("foo" + notString) as String
  }

  private fun testSafeCast(a: Any): Foo? {
    return a as? Foo
  }

  fun testCastToSameType() {
    val a = 1 as Int
    val b = true as Boolean
    val c = 'b' as Char
    val d = "Hello World" as String
    val f = Foo() as Foo
    val g: Int? = 1
    val h = g as Int?
  }

  fun testRedundantVarianceCasts(holder: Holder<Throwable>?) {
    val a = holder as Holder<Throwable>?
    val outA = holder as Holder<out Throwable>?
    val inA = holder as Holder<in Throwable>?

    val b = holder!! as Holder<Throwable>
    val outB = holder!! as Holder<out Throwable>
    val inB = holder!! as Holder<in Throwable>

    val c = holder as Holder<Throwable>
    val outC = holder as Holder<out Throwable>
    val inC = holder as Holder<in Throwable>
  }
}

class Holder<T>
